File: content/5-concurrency/1-async-basics/go/main.go
package main

import (
	"fmt"
	"sync"
	"time"
)

func slowTask(name string, delay time.Duration) string {
	time.Sleep(delay)
	return name + " done"
}

func main() {
	start := time.Now()

	var wg sync.WaitGroup
	wg.Add(2)

	results := make(chan string, 2)

	go func() {
		defer wg.Done()
		results <- slowTask("A", 600*time.Millisecond)
	}()

	go func() {
		defer wg.Done()
		results <- slowTask("B", 900*time.Millisecond)
	}()

	wg.Wait()
	close(results)

	for r := range results {
		fmt.Println(r)
	}

	fmt.Printf("elapsed=%dms\n", time.Since(start).Milliseconds())
}

File: content/5-concurrency/1-async-basics/python/main.py
import asyncio
import time


async def slow_task(name: str, delay_s: float) -> str:
    await asyncio.sleep(delay_s)
    return f"{name} done"


async def main() -> None:
    start = time.time()

    a, b = await asyncio.gather(
        slow_task("A", 0.6),
        slow_task("B", 0.9),
    )

    print(a)
    print(b)
    print(f"elapsed={int((time.time() - start) * 1000)}ms")


asyncio.run(main())

